R/location.R

```diff
@@ -447,48 +447,115 @@ orderly_location_pull_packet <- function(expr,
 ##'
 ##' @title Push tree to location
 ##'
-##' @param packet_id One or more packets to push to the server
+##' @param expr An expression to search for.  Often this will be a
+##'   vector of ids, but you can use a query here.
 ##'
 ##' @param location The name of a location to push to (see
 ##' [orderly2::orderly_location_list] for possible values).
 ##'
+##' @param dry_run Logical, indicating if we should print a summary
+##'   but not make any changes.
+##'
 ##' @inheritParams orderly_metadata
+##' @inheritParams orderly_search
 ##'
 ##' @return Invisibly, details on the information that was actually
 ##'   moved (which might be more or less than what was requested,
 ##'   depending on the dependencies of packets and what was already
 ##'   known on the other location).
 ##'
 ##' @export
-orderly_location_push <- function(packet_id, location, root = NULL) {
+orderly_location_push <- function(expr, location, name = NULL, dry_run = FALSE,
+                                  root = NULL) {
   root <- root_open(root, require_orderly = FALSE)
+  assert_scalar_logical(dry_run)
   location_name <- location_resolve_valid(location, root,
                                           include_local = FALSE,
                                           include_orphan = FALSE,
                                           allow_no_locations = FALSE,
                                           environment())
-  plan <- location_build_push_plan(packet_id, location_name, root)
+  if (expr_is_literal_id(expr, name)) {
+    ids <- expr
+    err <- setdiff(ids, root$index$unpacked())
+    if (length(err)) {
+      cli::cli_abort("Trying to push unknown packet{?s}: {squote(err)}")
+    }
+  } else {
+    ids <- orderly_search(expr, name = name, root = root)
+    if (length(ids) == 0) {
+      cli_alert_warning("Query returned no packets to push")
+    }
+  }
+
+  plan <- location_build_push_plan(ids, location_name, root)
+
+  if (length(plan$files) == 0 && length(plan$packet_id) == 0) {
+    cli_alert_success("Nothing to push, everything up to date")
+  } else {
+    cli_alert_info(
+      paste("Pushing {length(plan$files)} file{?s} for",
+            "{length(plan$packet_id)} packet{?s}"))
```

Comment:
is just length too little info? wonder if we can list these out with a ... at least, better yet list out an abbreviated version and let them expand if they wish

Reply:
We don't actually have filenames at this point because of the 1:many correspondence of hash to file.  All we hold there is hashes and ids neither of which are terribly useful